src/wallet/wallet.cpp

```diff
@@ -1542,9 +1542,12 @@ void CWallet::ReacceptWalletTransactions()
 bool CWalletTx::RelayWalletTransaction(CConnman* connman)
 {
     assert(pwallet->GetBroadcastTransactions());
-    if (!IsCoinBase())
+    if (!IsCoinBase() && !isAbandoned() && GetDepthInMainChain() == 0)
     {
-        if (GetDepthInMainChain() == 0 && !isAbandoned() && InMempool()) {
+        LOCK(mempool.cs);
+        CValidationState state;
+        /* GetDepthInMainChain already catches known conflicts. */
+        if (mempool.exists(GetHash()) || AcceptToMemoryPool(maxTxFee, state)) {
```

Comment:
If you use `InMempool()` here instead of `mempool.exists(GetHash())`, you can drop the `LOCK(mempool.cs)`.